pennylane/devices/default_qubit_jax.py

```diff
@@ -0,0 +1,193 @@
+    _gather = staticmethod(lambda array, indices: array[indices])
+    _einsum = staticmethod(jnp.einsum)
+    _cast = staticmethod(jnp.array)
+    _transpose = staticmethod(jnp.transpose)
+    _tensordot = staticmethod(
+        lambda a, b, axes: jnp.tensordot(
+            a, b, axes if isinstance(axes, int) else list(map(tuple, axes))
+        )
+    )
+    _conj = staticmethod(jnp.conj)
+    _imag = staticmethod(jnp.imag)
+    _roll = staticmethod(jnp.roll)
+    _stack = staticmethod(jnp.stack)
+
+    def __init__(self, wires, *, shots=1000, analytic=True, prng_key=None):
+        super().__init__(wires, shots=shots, analytic=analytic, cache=0)
+
+        # prevent using special apply methods for these gates due to slowdown in jax
+        # implementation
+        del self._apply_ops["PauliY"]
+        del self._apply_ops["Hadamard"]
+        del self._apply_ops["CZ"]
+        self._prng_key = prng_key
+
+    @classmethod
+    def capabilities(cls):
+        capabilities = super().capabilities().copy()
+        capabilities.update(
+            passthru_interface="jax",
+            supports_reversible_diff=False,
+        )
+        return capabilities
+
+    @staticmethod
+    def _scatter(indices, array, new_dimensions):
+        new_array = jnp.zeros(new_dimensions, dtype=array.dtype.type)
+        new_array = new_array.at[indices].set(array)
+        return new_array
+
+    def _get_unitary_matrix(self, unitary):
+        """Return the matrix representing a unitary operation.
+
+        Args:
+            unitary (~.Operation): a PennyLane unitary operation
+
+        Returns:
+            array[complex]: Returns a 2D matrix representation of
+            the unitary in the computational basis, or, in the case of a diagonal unitary,
+            a 1D array representing the matrix diagonal.
+        """
+        op_name = unitary.name
+        if op_name in self.parametric_ops:
+            if op_name == "MultiRZ":
+                return self.parametric_ops[unitary.name](*unitary.parameters, len(unitary.wires))
+            return self.parametric_ops[unitary.name](*unitary.parameters)
+
+        if isinstance(unitary, DiagonalOperation):
+            return unitary.eigvals
+
+        return unitary.matrix
+
+    def sample_basis_states(self, number_of_states, state_probability):
+        """Sample from the computational basis states based on the state
+        probability.
+
+        This is an auxiliary method to the generate_samples method.
+
+        Args:
+            number_of_states (int): the number of basis states to sample from
+
+        Returns:
+            List[int]: the sampled basis states
+        """
+        if self._prng_key is None:
+            # Assuming op-by-op, so we'll just make one.
+            key = jax.random.PRNGKey(np.random.randint(0, 2 ** 31))
+        else:
+            key = self._prng_key
+        return jax.random.choice(key, number_of_states, shape=(self.shots,), p=state_probability)

```

Comment:
Such a small file for such a big feature.

Reply:
That just points to how well designed pennylane is!

Reply:
Amazing!